video/gazebocamerainput.h

```diff
@@ -0,0 +1,133 @@
+
+// OpenCV
+#include <opencv2/opencv.hpp>
+
+// Standard C++ includes
+#include <stdexcept>
+#include <mutex>
+#include <atomic>
+
+namespace BoBRobotics {
+namespace Video {
+
+constexpr const char *GazeboCameraDeviceName = "gazebo_camera";
+
+//----------------------------------------------------------------------------
+// BoBRobotics::Video::GazeboCameraInput
+//----------------------------------------------------------------------------
+//! A thin wrapper for reading from any Gazebo camera feed
+class GazeboCameraInput : public Input
+{
+public:
+    /*!
+     * \brief Create a video stream using a new transport node and default topic
+     */
+    GazeboCameraInput()
+    : GazeboCameraInput("/gazebo/default/camera/link/camera/image")
+    {}
+    /*!
+     * \brief Create a video stream using a new transport node
+     */
+    GazeboCameraInput(const std::string &topic, const std::string &cameraName = GazeboCameraDeviceName)
+    : m_CameraTopic(topic), m_CameraName(cameraName)
+    {
+        // Load gazebo as a client
+        #if GAZEBO_MAJOR_VERSION < 6
+        gazebo::setupClient(0, 0);
+        #else
+        gazebo::client::setup(0, 0);
+        #endif
+        //Setup image subscription node
+        m_ImageNode= gazebo::transport::NodePtr(new gazebo::transport::Node());
+        m_ImageNode->Init();
+        subscribeToGazeboCamera();
+    }
+    /*!
+     * \brief Create a video stream for a Gazebo topic using a user-provided tranport node
+     *
+     * @param node Gazebo transport node
+     * @param topic Gazebo transport topic on which to subscribe
+     */
+    
+    GazeboCameraInput(gazebo::transport::NodePtr node, const std::string &topic="/gazebo/default/camera/link/camera/image", const std::string &cameraName = GazeboCameraDeviceName)
+      : m_CameraTopic(topic), m_CameraName(cameraName), m_ImageNode(node)
+    {
+        subscribeToGazeboCamera();
+    }
+
+    void subscribeToGazeboCamera(){
+        // Subscribe to the topic, and register a callback
+        m_HaveReceivedFrames = false;
+        m_ImageSub = m_ImageNode->Subscribe(m_CameraTopic, &GazeboCameraInput::OnImageMsg, this);
+        LOG_INFO << "Subsribed to "<< m_CameraTopic <<"\n";
+    }
+    //------------------------------------------------------------------------
+    // Video::Input virtuals
+    //------------------------------------------------------------------------
+    virtual std::string getCameraName() const override
+    {
+        return "gazebo_camera";
+    }
+
+    virtual cv::Size getOutputSize() const override
+    {
+        return m_OutSize;
+    }
+
+    virtual bool readFrame(cv::Mat &outFrame) override
+    {
+        if(!m_HaveReceivedFrames.load())
+            return false;
+        std::lock_guard<std::mutex> lck(m_Mtx);
+        outFrame = m_ReceivedImage;
+        // If there's no error, then we have updated frame and so return true
+        return true;
+    }
+
+
+private:
+    const std::string m_CameraTopic;
```

Comment:
Do you need to store this string once it's used?

Reply:
Not really. We can remove the member by passing the string in the subscribeToGazeboCamera() function call

Reply:
I think that would be cleaner. No need to copy the string if we're only using it once.